Register a new Session

sgomez opened the pull request:
Closes #11

sgomez committed:
Create Session Model

sgomez committed:
Implement SessionRepositoryInterface

- Use cache component to save the session.
- Added a normalizer to convert the Session Model to a serializable string.
- Change the phpstan rules to avoid a issue with NormalizerInterface phpdoc.

sgomez committed:
Register new sessions

Added a functional test to check  the handler works with the CacheSessionRepository service.

MKodde commented on src/Domain/Session/Model/SessionExpiryDate.php:
```diff
@@ -0,0 +1,40 @@
+
+    public static function fromISOString(string $expiryDateISO): SessionExpiryDate
+    {
+        return new self(new DateTimeImmutable($expiryDateISO));
```
Here you might want to verify the ISO string is valid. For example, by creating an instance in a try catch block. That way you can present a controlled exception message if the string is not valid.

Leaving it like this is also okay with me. I personally like to have controll over the messages ending up in the logs

MKodde commented on src/Domain/Session/Model/SessionExpiryDate.php:
```diff
@@ -0,0 +1,40 @@
+
+    public function equals(SessionExpiryDate $other): bool
+    {
+        return $this->value->getTimestamp() === $other->value->getTimestamp();
```
Did you know you can compare DateTime (Immutable) instances with comparison operators? Possible as of php 5.2

```
return $this->value->getTimestamp() === $other->value->getTimestamp();
// could be shortened to:
return $this->value == $other->value;
```

https://www.php.net/manual/en/datetime.diff.php

sgomez replied:
Yes, I know. But phpcs requires to use strict equal operator (===).

MKodde commented on support/qa/phpstan.neon:
```diff
@@ -5,3 +5,14 @@ parameters:
         - ../../src
+    ignoreErrors:
+        -
+            message: '#Method [a-zA-Z0-9\\_]+::[a-zA-Z0-9]+\(\) has parameter \$context#'
+            path: '**/Infrastructure/Serializer/SessionNormalizer.php'
+        -
+            message: '#Method [a-zA-Z0-9\\_]+::[a-zA-Z0-9]+\(\) return type has no value type specified in iterable type array#'
+            path: '**/Infrastructure/Serializer/SessionNormalizer.php'
+        -
+            message: '#Method [a-zA-Z0-9\\_]+::[a-zA-Z0-9]+\(\) return type with generic class ArrayObject does not specify its types: TKey, TValue#'
+            path: '**/Infrastructure/Serializer/SessionNormalizer.php'
```
Wouldn't these exceptions be easily removed if you'd write the docblock (currently inherited from parent) yourself?

sgomez replied:
AFAIK, the phpdoc can be extended, not overwrited. I waste a lot of time trying to avoid this issue adding a local `@param`. But inherit docblock is wrong and I haven't found a way to avoid it.

I can not remove `inheritdoc` tag, because then phpcs and phpmd fails too. Is this or add annotations to avoid phpcs and phpmd errors.

Maybe using psalm instead phpstan...

MKodde replied:
If its up to me, just leave the exceptions in place. Let's stay alert if this comes back more often. If it does we can always switch to Psalm, or configure Stan in a less strict manner.

MKodde reviewed (approved):
Great work!

sgomez committed:
Fix SessionNormalizer methods signatures

- supportsNormalization has an extra parameter
- added array return type to normalize method, this removes an error in phpstan

sgomez committed:
Check valid session expiry date ISO string
